q2_fmt/tests/test_engraftment.py

```diff
@@ -1250,10 +1254,336 @@ def test_rename_features_with_blank_label(self):
+                            iterations=999)
+
+    def test_create_mismatched_pairs(self):
+        metadata_df = pd.DataFrame({
+            'id': ['sample1', 'sample2', 'sample3',
+                   'donor1', 'donor2', 'donor3'],
+            'Ref': ['donor1', 'donor2', 'donor3', np.nan, np.nan,
+                    np.nan],
+            'subject': ['sub1', 'sub2', 'sub3', np.nan, np.nan,
+                        np.nan],
+            'group': [1, 1, 1, np.nan, np.nan,
+                      np.nan],
+            "Location": [np.nan, np.nan,
+                         np.nan, 'test', 'test',
+                         'test']}).set_index('id')
+
+        recip_df = pd.DataFrame({
+            'id': ['sample1', 'sample2', 'sample3'],
+            'Feature1': [1, 0, 0],
+            'Feature2': [0, 1, 0],
+            'Feature3': [0, 0, 1]}).set_index('id')
+
+        used_references = pd.Series(data=['donor1', 'donor2', 'donor3'],
+                                    index=['sample1', 'sample2', 'sample3'],
+                                    name='Ref')
+        used_references.index.name = "id"
+        mismatched_df = _create_mismatched_pairs(recip_df, metadata_df,
+                                                 used_references,
+                                                 reference_column='Ref')
+        exp_mismatched_df = pd.DataFrame({'id': ["sample1", "sample1",
+                                                 "sample2", "sample2",
+                                                 "sample3", "sample3"],
+                                          "Ref": ["donor2", "donor3",
+                                                  "donor1", "donor3",
+                                                  "donor1", "donor2"]}
+                                         ).set_index('id')
+        pd.testing.assert_frame_equal(mismatched_df, exp_mismatched_df)
+
+    def test_mask_recipient(self):
+        donor_mask = np.array([[0, 1, 0], [0, 0, 1], [1, 0, 0]])
+        recip_df = pd.DataFrame({
+            'id': ['sample1', 'sample2', 'sample3'],
+            'Feature1': [1, 0, 0],
+            'Feature2': [0, 1, 0],
+            'Feature3': [0, 0, 1]}).set_index('id')
+        recip_mask = _mask_recipient(donor_mask, recip_df)
+        exp_r_mask = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
+        np.testing.assert_array_equal(recip_mask, exp_r_mask)
+
+    def test_mask_recipient_donor_one(self):
+        donor_mask = np.array([[0, 1, 0], [0, 1, 0], [0, 1, 0]])
+        recip_df = pd.DataFrame({
+            'id': ['sample1', 'sample2', 'sample3'],
+            'Feature1': [1, 0, 0],
+            'Feature2': [0, 1, 0],
+            'Feature3': [0, 0, 1]}).set_index('id')
+        recip_mask = _mask_recipient(donor_mask, recip_df)
+        exp_r_mask = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
+        np.testing.assert_array_equal(recip_mask, exp_r_mask)
+
+    def test_simulate_uniform_distro(self):
+        mismatch_peds = [0, 0, 0, 0, 0, 0]
```

Comment:
I recommend adding some diversity to this test. For example, make mismatch_peds = [1, 2, 3], and then confirm that you see each of those values at least once in the results (intermittent failure is possible, so it's worth mentioning that in a comment, noting that it should be extremely rare).